### PR TITLE
Typing support

### PR DESCRIPTION
**Pull request by palfrey**

### CODE REVIEW

File: asttokens/util.py [hunk 1]
```diff
@@ -59,21 +68,24 @@ def expect_token(token, tok_type, tok_str=None):
 
 # These were previously defined in tokenize.py and distinguishable by being greater than
 # token.N_TOKEN. As of python3.7, they are in token.py, and we check for them explicitly.
-if hasattr(token, 'ENCODING'):
+if sys.version_info >= (3, 7):
```
Comment (palfrey) [inline asttokens/util.py:71]: Because of https://mypy.readthedocs.io/en/stable/common_issues.html#python-version-and-system-platform-checks

File: tests/test_asttokens.py [hunk 2]
```diff
@@ -80,6 +80,8 @@ def test_token_methods(self):
     self.assertTrue(token.ISEOF(atok.find_token(atok.tokens[5], tokenize.NL).type))
 
   def test_unicode_offsets(self):
+    if six.PY2:
+      pytest.skip("Unicode parsing is different on Python 2")
```
Comment (palfrey) [inline tests/test_asttokens.py:84]: No idea how anyone got this working on 2.7. This should have failed before now! See also https://github.com/gristlabs/asttokens/runs/4493079453?check_suite_focus=true

File: tests/test_mark_tokens.py [hunk 3]
```diff
@@ -560,6 +561,8 @@ def test_statements_with_semicolons(self):
 
 
   def test_complex_numbers(self):
+    if six.PY2:
+      pytest.skip("Complex number parsing is different on Python 2")
```
Comment (palfrey) [inline tests/test_mark_tokens.py:565]: Similar 2.7 failure to the one above

**Comment by palfrey** [comment]
Any thoughts on this @dsagal @alexmojaki?

File: asttokens/asttokens.py [hunk 4]
```diff
@@ -43,14 +43,14 @@ class ASTTokens(object):
...
+    # type: (str, bool, Optional[Module], str) -> None
     self._filename = filename
     self._tree = ast.parse(source_text, filename) if parse else tree
 
     # Decode source after parsing to let Python 2 handle coding declarations.
     # (If the encoding was not utf-8 compatible, then even if it parses correctly,
     # we'll fail with a unicode error here.)
-    if isinstance(source_text, six.binary_type):
-      source_text = source_text.decode('utf8')
+    source_text = six.ensure_str(source_text)
```
Comment (alexmojaki) [inline asttokens/asttokens.py:53]: Should probably be `ensure_text`

File: asttokens/asttokens.py [hunk 5]
```diff
@@ -66,56 +66,65 @@ def __init__(self, source_text, parse=False, tree=None, filename='<unknown>'):
...
 
   def _generate_tokens(self, text):
+    # type: (str) -> Iterator[Token]
     """
     Generates tokens for the given code.
     """
     # This is technically an undocumented API for Python3, but allows us to use the same API as for
     # Python2. See http://stackoverflow.com/a/4952291/328565.
-    for index, tok in enumerate(tokenize.generate_tokens(io.StringIO(text).readline)):
+    for index, tok in enumerate(tokenize.generate_tokens(six.StringIO(text).readline)):
```
Comment (alexmojaki) [inline asttokens/asttokens.py:88]: Since you're having Python 2 unicode issues, can you revert this and see if it helps?

File: asttokens/asttokens.py [hunk 6]
```diff
@@ -175,13 +193,15 @@ def token_range(self, first_token, last_token, include_extra=False):
         yield self._tokens[i]
 
   def get_tokens(self, node, include_extra=False):
+    # type: (NodeNG, bool) -> Iterator
```
Comment (alexmojaki) [inline asttokens/asttokens.py:196]:
```suggestion
    # type: (NodeNG, bool) -> Iterator[Token]
```

File: asttokens/mark_tokens.py [hunk 7]
```diff
@@ -147,11 +157,13 @@ def _expand_to_matching_pairs(self, first_token, last_token, node):
   # that will actually be assigned.
 
   def visit_default(self, node, first_token, last_token):
+    # type: (NodeNG, util.Token, util.Token) -> Tuple[util.Token, util.Token]
     # pylint: disable=no-self-use
     # By default, we don't need to adjust the token we computed earlier.
     return (first_token, last_token)
 
   def handle_comp(self, open_brace, node, first_token, last_token):
+    # type: (NodeNG, Optional[str], util.Token, util.Token) -> Tuple[util.Token, util.Token]
```
Comment (alexmojaki) [inline asttokens/mark_tokens.py:166]:
```suggestion
    # type: (str, NodeNG, util.Token, util.Token) -> Tuple[util.Token, util.Token]
```

File: asttokens/mark_tokens.py [hunk 8]
```diff
@@ -162,28 +174,38 @@ def handle_comp(self, open_brace, node, first_token, last_token):
...
+      # type: (NodeNG, util.Token, util.Token) -> Tuple[util.Token, util.Token]
       return self.handle_comp('{', node, first_token, last_token)
 
     def visit_dictcomp(self, node, first_token, last_token):
+      # type: (NodeNG, util.Token, util.Token) -> Tuple[util.Token, util.Token]
       return self.handle_comp('{', node, first_token, last_token)
 
-  def visit_comprehension(self, node, first_token, last_token):
+  def visit_comprehension(self,
+                          node,  # type: util.Token
```
Comment (alexmojaki) [inline asttokens/mark_tokens.py:191]:
```suggestion
                          node,  # type: NodeNG
```

Is it possible to make a type alias to apply to all these methods?

File: asttokens/mark_tokens.py [hunk 9]
```diff
@@ -232,11 +257,17 @@ def visit_call(self, node, first_token, last_token):
       first_token = self._code.next_token(first_token)
     return (first_token, last_token)
 
-  def visit_subscript(self, node, first_token, last_token):
+  def visit_subscript(self,
+                      node,  # type: util.Token
```
Comment (alexmojaki) [inline asttokens/mark_tokens.py:261]:
```suggestion
                      node,  # type: NodeNG
```

File: asttokens/mark_tokens.py [hunk 10]
```diff
@@ -232,11 +257,17 @@ def visit_call(self, node, first_token, last_token):
...
+                      first_token,  # type: util.Token
+                      last_token,  # type: util.Token
+                      ):
+    # type: (...) -> Tuple[util.Token, util.Token]
     last_token = self.handle_following_brackets(node, last_token, '[')
     return (first_token, last_token)
 
   def visit_slice(self, node, first_token, last_token):
+    # type: (util.Token, util.Token, util.Token) -> Tuple[util.Token, util.Token]
```
Comment (alexmojaki) [inline asttokens/mark_tokens.py:270]:
```suggestion
    # type: (NodeNG, util.Token, util.Token) -> Tuple[util.Token, util.Token]
```

File: asttokens/mark_tokens.py [hunk 11]
```diff
@@ -256,6 +287,7 @@ def visit_slice(self, node, first_token, last_token):
     return (first_token, last_token)
 
   def handle_bare_tuple(self, node, first_token, last_token):
+    # type: (util.Token, util.Token, util.Token) -> Tuple[util.Token, util.Token]
```
Comment (alexmojaki) [inline asttokens/mark_tokens.py:290]:
```suggestion
    # type: (NodeNG, util.Token, util.Token) -> Tuple[util.Token, util.Token]
```

File: asttokens/util.py [hunk 12]
```diff
@@ -201,12 +222,13 @@ def visit_tree(node, previsit, postvisit):
...
         stack.insert(ins, (n, pvalue, _PREVISIT))
     else:
-      ret = postvisit(current, par_value, value)
+      ret = postvisit(current, par_value, cast(Optional[Token], value))
   return ret
 
 
 
 def walk(node):
+  # type: (Module) -> Iterator[Module]
```
Comment (alexmojaki) [inline asttokens/util.py:231]:
```suggestion
  # type: (Module) -> Iterator[NodeNG]
```

File: asttokens/mark_tokens.py (hunk 8 above)
Comment (palfrey) [reply asttokens/mark_tokens.py:191]: Maybe. Not sure how!

File: asttokens/util.py (hunk 12 above)
Comment (palfrey) [reply asttokens/util.py:231]: I don't think this is right, given the first item it returns is the input Module. Possibly `Union[Module, NodeNG]` but hard to tell in the full-on dynamic typing going on here.

File: asttokens/asttokens.py (hunk 4 above)
Comment (palfrey) [reply asttokens/asttokens.py]: Tried this, and it breaks the Python 2 typing

File: asttokens/util.py (hunk 12 above)
Comment (alexmojaki) [reply asttokens/util.py:231]: Most of the mentions of `NodeNG` aren't right since the functions work with either an `astroid` or `ast` node. I thought you were just using `NodeNG` to mean `ast.AST` for some reason. But this definitely yields more than just modules.

**Comment by alexmojaki** [comment]
Can you revert the skips you added for Python 2 and see if they pass now?

**Comment by palfrey** [comment]
> Can you revert the skips you added for Python 2 and see if they pass now?

I've just tried that locally, and they're still broken

**Comment by palfrey** [comment]
> Please revert the skips and the unicode related changes and push. The tests were working before, and only adding type comments shouldn't change that. If the previous code broke typing, that's a typing problem and that's secondary.

Yup, you're right. I reset bits of the changes, removed the test skips and added a typing workaround [for an upstream bug](https://github.com/python/typeshed/issues/7003) and it's working better now. Still one test to fix, but working on it...

File: asttokens/asttokens.py (hunk 5 above)
Comment (palfrey) [reply asttokens/asttokens.py]: Resolved. [Upstream typing issue](https://github.com/python/typeshed/issues/7003) and there's now a `cast` to fix the type as a workaround.

File: asttokens/asttokens.py (hunk 4 above)
Comment (palfrey) [reply asttokens/asttokens.py]: Yup, you were right. Done now.

File: asttokens/mark_tokens.py [hunk 13]
```diff
@@ -331,11 +381,13 @@ def handle_num(self, node, value, first_token, last_token):
     return (first_token, last_token)
 
   def visit_num(self, node, first_token, last_token):
+    # type: (NodeNG, util.Token, util.Token) -> Tuple[util.Token, util.Token]
     return self.handle_num(node, node.n, first_token, last_token)
 
   # In Astroid, the Num and Str nodes are replaced by Const.
   def visit_const(self, node, first_token, last_token):
-    if isinstance(node.value, numbers.Number):
```
Comment (alexmojaki) [inline asttokens/mark_tokens.py:338]: This is probably the source of the complex number test failure...

Comment (palfrey) [reply asttokens/mark_tokens.py:338]: And solved by c0bdb710fe790b73bcad3cef4629119fb075b989

File: asttokens/util.py (hunk 12 above)
Comment (palfrey) [reply asttokens/util.py:231]: I've just ran some experiments on this one and found the correct set of types was `Union[Module, AST, NodeNG]`

Are you saying that most of the `NodeNG` refs should actually be `Union[AST, NodeNG]`?

Comment (alexmojaki) [reply asttokens/util.py]: Sorry I forgot about this. Yes, technically, I think they should usually be `Union[AST, NodeNG]`, i.e. they work with both `ast` and `astroid`.

Comment (palfrey) [reply asttokens/util.py]: I've added a `AstNode` type alias for `Union[AST, NodeNG]` and moved over most of the util usages of `NodeNG` to that. There are two exceptions, because they use features only on `NodeNG`

File: asttokens/util.py [hunk 14]
```diff
@@ -124,30 +140,36 @@ def iter_children_ast(node):
...
 
 
 def is_starred(node):
+  # type: (AstNode) -> bool
   """Returns whether node is a starred expression node."""
   return node.__class__.__name__ == 'Starred'
 
 
 def is_slice(node):
+  # type: (NodeNG) -> bool
```
Comment (palfrey) [inline asttokens/util.py:172]: `.elts` is only on `NodeNG`, so can't use `ASTNode`

Comment (alexmojaki) [reply asttokens/util.py:172]: It definitely exists in `ast`:

```
>>> ast.parse('1, 2').body[0].value.elts
[<_ast.Constant object at 0x7f56dd8ba5b0>, <_ast.Constant object at 0x7f56dd8ba3d0>]
```

Maybe you need to specify specific types of node?

File: setup.cfg [hunk 15]
```diff
@@ -36,7 +36,13 @@ install_requires = six
 setup_requires = setuptools>=44; setuptools_scm[toml]>=3.4.3
 
 [options.extras_require]
-test = astroid<=2.5.3; pytest
+test = 
+    astroid<=2.5.3
+    pytest
+    typing; python_version < "3.5"
```
Comment (alexmojaki) [inline setup.cfg:42]: Isn't this an essential (i.e. not extra) requirement for Python 2 users?

File: asttokens/mark_tokens.py [hunk 16]
```diff
@@ -61,14 +68,15 @@ def _visit_before_children(self, node, parent_token):
     return (token or parent_token, token)
 
   def _visit_after_children(self, node, parent_token, token):
+    # type: (NodeNG, Optional[util.Token], Optional[util.Token]) -> None
```
Comment (alexmojaki) [inline asttokens/mark_tokens.py:71]: There's still lots of `NodeNG` here and in `asttokens.py` that should really be `AstNode`. Note that the library adds several attributes like `first_token` to the nodes, you'll have to find a way to deal with that.

File: setup.cfg (hunk 15 above)
Comment (palfrey) [reply setup.cfg]: Yup! Fixed

File: asttokens/mark_tokens.py (hunk 16 above)
Comment (palfrey) [reply asttokens/mark_tokens.py]: I've replaced them with AstNode, and fixed the definition of that to use an `EnhancedAST` types marker class that basically just goes "we promise ASTs have some extra fields"

File: asttokens/asttokens.py [hunk 17]
```diff
@@ -175,13 +198,15 @@ def token_range(self, first_token, last_token, include_extra=False):
         yield self._tokens[i]
 
   def get_tokens(self, node, include_extra=False):
+    # type: (AstNode, bool) -> Iterator[Token]
     """
     Yields all tokens making up the given node. If include_extra is True, includes non-coding
     tokens such as tokenize.NL and .COMMENT.
     """
-    return self.token_range(node.first_token, node.last_token, include_extra=include_extra)
+    return self.token_range(cast(Token, node.first_token), cast(Token, node.last_token), include_extra=include_extra)
```
Comment (alexmojaki) [inline asttokens/asttokens.py:206]: Doesn't `EnhancedAST` remove the need for these casts?

File: asttokens/mark_tokens.py [hunk 18]
```diff
@@ -195,9 +219,10 @@ def handle_attr(self, node, first_token, last_token):
   visit_delattr = handle_attr
 
   def handle_def(self, node, first_token, last_token):
+    # type: (AstNode, util.Token, util.Token) -> Tuple[util.Token, util.Token]
     # With astroid, nodes that start with a doc-string can have an empty body, in which case we
     # need to adjust the last token to include the doc string.
-    if not node.body and getattr(node, 'doc', None):
+    if not getattr(node, 'body', None) and getattr(node, 'doc', None):
```
Comment (alexmojaki) [inline asttokens/mark_tokens.py:225]: Why this change?

Comment (palfrey) [reply asttokens/mark_tokens.py]: This was mostly because it predated the work with narrowing the types for the various handle/visit functions down to the ones they were actually using. I tried doing that here, and the list of types is pretty large (I got to about at least 6 triggered by the tests), and they're annoyingly lacking in common parent classes to use instead. I've reset this back to it's original code and added a `type: ignore` comment.

File: asttokens/asttokens.py (hunk 17 above)
Comment (palfrey) [reply asttokens/asttokens.py]: Yup, removed.

**Comment by alexmojaki** [comment]
:+1: I'm happy. Thanks for your patience! @dsagal ?

**Comment by alexmojaki** [comment]
Thanks for all your hard work @palfrey !
